[SRVOCF-288] Add on-cluster build docs (#45722)
openshift/openshift-docs: serverless/functions/serverless-functions-on-cluster-builds.adoc

> :_content-type: PROCEDURE

Reviewer: This should be `:_content-type: ASSEMBLY` for an assembly.

> [id="serverless-functions-on-cluster-builds"]

Reviewer: This anchor is correct without the `_{context}`, since it's the assembly's anchor. Only modules need to include the _{context}.

= Building and deploying functions on-cluster
:context: serverless-functions-on-cluster-builds
include::_attributes/common-attributes.adoc[]

toc::[]

You can use the Knative (`kn`) CLI to initiate a function project build directly on the cluster. To build a function project in this way, the source code for your function project must exist in a Git repository branch that is accessible to your cluster.

:FeatureName: {FunctionsProductName}
include::snippets/technology-preview.adoc[leveloffset=+1]

[id="prerequisites_serverless-functions-on-cluster-builds"]
== Prerequisites

> * xref:../../cicd/pipelines/installing-pipelines.adoc[{pipelines-title}] must be installed on your cluster.

Reviewer: +1 to what John says. Technically this is in an assembly now, but this procedure really does need to be moved into its own procedure module

Also, the xref is missing an #anchor

Author: Removed the xref anyway when moving to module

> * xref:../../cicd/pipelines/installing-pipelines.adoc[{pipelines-title}] must be installed on your cluster.

Author: @bergerhoffer @johnwilkins I believe this was why it was done as an assembly rather than a procedure. It's a much messier UX to say you need pipelines as a prereq but then have a user have to "see additional resources" to find instructions on that, rather than just being able to include the xref in the prereqs. I would prefer to use the prereqs like [this](https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#anchoring-prerequisites-additional-resources-and-next-steps-titles-in-assemblies) but then it wouldn't be using the **.Prerequisites** in the module, unless I repeat them there.

* You have installed the OpenShift (`oc`) CLI.

* You have installed the Knative (`kn`) CLI.

[id="serverless-functions-on-cluster-builds-procedure"]
== Procedure

. In each namespace where you want to run {pipelines-shortname} and deploy a function, you must create the following resources:

.. Create the functions buildpacks Tekton task to be able to build the function image:
+
[source,terminal]
----

> $ oc apply -f https://raw.githubusercontent.com/openshift-knative/kn-plugin-func/tree/serverless-1.22.0/pipelines/resources/tekton/task/func-buildpacks/0.1/func-buildpacks.yaml

Reviewer 2: one last thing (here without the `tree`) my bad.. sorry for that...
```suggestion
$ oc apply -f https://raw.githubusercontent.com/openshift-knative/kn-plugin-func/serverless-1.22.0/pipelines/resources/tekton/task/func-buildpacks/0.1/func-buildpacks.yaml
```

----

.. Create the `kn func` deploy Tekton task to be able to deploy the function in the pipeline:
+
[source,terminal]
----

> $ oc apply -f https://raw.githubusercontent.com/openshift-knative/kn-plugin-func/tree/serverless-1.22.0/pipelines/resources/tekton/task/func-deploy/0.1/func-deploy.yaml

Reviewer 2: and same here
```suggestion
$ oc apply -f https://raw.githubusercontent.com/openshift-knative/kn-plugin-func/serverless-1.22.0/pipelines/resources/tekton/task/func-deploy/0.1/func-deploy.yaml
```

----

. Create a function:
+
[source,terminal]
----
$ kn func create <function_name> -l <runtime>
----

. After you have created a new function project, you must add the project to a Git repository and ensure that the repository is available to the cluster. Information about this Git repository is used to update the `func.yaml` file in the next step.

. Update the configuration in the `func.yaml` file for your function project to enable on cluster builds for the Git repository:
+
[source,yaml]
----
...
build: git <1>
git:
  url: <git_repository_url> <2>
  revision: main <3>
  contextDir: <directory_path> <4>
...
----
<1> Required. Specify `git` build type.
<2> Required. Specify the Git repository that contains your function's source code.
<3> Optional. Specify the Git repository revision to be used. This can be a branch, tag or commit.

> <4> Optional. Specify this path if the function is not located in the Git repository root folder.

Reviewer 3: Consider replacing "this path" with "the directory path" or "the function's directory path"

:_content-type: PROCEDURE
[id="serverless-functions-on-cluster-builds"]
= Building and deploying functions on-cluster
:context: serverless-functions-on-cluster-builds
include::_attributes/common-attributes.adoc[]

toc::[]

> You can use the Knative (`kn`) CLI to push a function project directly to the cluster, which triggers a build and deployment of the function. You can also push a function project to a source code repository branch, which triggers a build of the function with the updated changes and deploys the updated function on the cluster.

Reviewer 4: This wording is a little imprecise. The CLI does not actually push a project to the cluster. It instead only creates a Pipeline on the cluster which pulls the function project source from GitHub. Pushing code directly to the cluster is a desired capability but not implemented yet. It is tracked here: https://github.com/knative-sandbox/kn-plugin-func/issues/952 

Perhaps rephrasing to something like this would be clearer.
```suggestion
You can use the Knative (`kn`) CLI to initiate a function project build directly on the cluster. Building a function project in this way requires that your function project source code exist in a Git repository branch accessible to the cluster.
```

:FeatureName: {FunctionsProductName}
include::snippets/technology-preview.adoc[leveloffset=+1]

[id="prerequisites_serverless-functions-on-cluster-builds"]
== Prerequisites

* xref:../../cicd/pipelines/installing-pipelines.adoc[{pipelines-title}] must be installed on your cluster.

* You have installed the OpenShift (`oc`) CLI.

* You have installed the Knative (`kn`) CLI.

[id="serverless-functions-on-cluster-builds-procedure"]
== Procedure

. In each namespace where you want to run {pipelines-shortname} and deploy a function, you must create the following resources:

.. Create the functions buildpacks Tekton task to be able to build the function image:

> +

Author: I believe this is the one I had asked @lance about and he said that it's installed by Pipelines by default. We can't have links to upstream git repos, so if this needs to be included it would need to be available in the `openshift-knative` repo instead of the `tekton` one. @lance PTAL.

> +

Reviewer 4: @abrennan89 is correct that this Task should be already installed with OpenShift Pipelines. This should be confirmed in QA @jrangelramos.

> +

Reviewer 2: Hey @lance .. yea I confirmed that. It looks like that task is not installed by default by Openshift Pipelines. It fails with me on OCP, so I had to install it to get it working.

> +

Reviewer 4: @jrangelramos OK good to know. Perhaps it's part of the catalog. Checking... 

Yes - https://github.com/tektoncd/catalog/tree/main/task/git-clone

This appears to manifest in the OpenShift Developer's Console when creating a pipeline. When you click the "Add task" button, you can choose the `git-clone` task. Here's a screen capture. On this cluster, it appears that the `git-clone` task from Red Hat is already installed.

![image](https://user-images.githubusercontent.com/15952/170284039-335da3f2-8326-4d7e-83ff-3b5ba7d01aa4.png)

There is community version that is not installed here.

![image](https://user-images.githubusercontent.com/15952/170283254-e235b515-2bd8-4ea6-a34d-cb54eb94a8cf.png)

You can install the community version by creating a pipeline which uses it, or directly from the catalog here https://hub.tekton.dev/tekton/Task/git-clone/0.6.

I didn't create the cluster, so it's possible that someone else manually installed the Red Hat `Task` manually by creating a pipeline. Or there may be some way to install a `Task` from the catalog without having to create a pipeline. I'm not sure.

I don't know the best way to document this. Is there any precedent here @abrennan89? Have we documented installing and specific `Task` dependencies yet?

> +

Author: No we haven't. I believe the correct thing to do would be to make sure it's documented in the Pipelines docs https://docs.openshift.com/container-platform/4.10/cicd/pipelines/working-with-pipelines-using-the-developer-perspective.html but it's probably a question for @Preeticp

> +

Reviewer 5: Hi @piyush-garg, @nikhil-thomas it would be great if you could provide some clarity wrt to the following questions: 

1. Is there a downstream equivalent for this: https://raw.githubusercontent.com/tektoncd/catalog/master/task/git-clone/0.4/git-clone.yaml? 
2. Is the tekton task git-clone installed by default when you install the Pipelines Operator?

> +

Reviewer 5: WRT:

> No we haven't. I believe the correct thing to do would be to make sure it's documented in the Pipelines docs https://docs.openshift.com/container-platform/4.10/cicd/pipelines/working-with-pipelines-using-the-developer-perspective.html but it's probably a question for @Preeticp

I believe it is documented in step 2 B here: https://docs.openshift.com/container-platform/4.10/cicd/pipelines/working-with-pipelines-using-the-developer-perspective.html#op-constructing-pipelines-using-pipeline-builder_working-with-pipelines-using-the-developer-perspective

> +

Reviewer 6: 1. We don't have any downstream equivalent for the git-clone task, we ship the same upstream task in the operator and fetched from upstream catalog https://github.com/tektoncd/operator/blob/main/hack/openshift/fetch-tektoncd-catalog-tasks.sh#L47 However the git-init image which gets used is from redhat, but it uses the same binary from upstream for git-init and just based on Ubi.

2. This task gets installed with openshift pipelines operator given the clustertasks are enabled in tekton config https://github.com/tektoncd/operator/blob/main/docs/TektonConfig.md#addon

> +

Reviewer 6: One more thing is this task get installed by operator as clustertask so it will be available in all namespaces

> +

Reviewer 5: @jrangelramos would it be possible for you to share the Pipeline with @piyush-garg, please? That would help him dig a bit deeper to resolve the issue. Thank you!

> +

Author: Additional question - does anyone know what OCP version added this task for the Operator? Our docs for serverless are supported on 4.6+, but I wonder if that's also the case for this Tekton part.

> +

Reviewer 4: @jrangelramos thanks for that update - I understand now. Where do we stand with docs?

[source,terminal]
----
$ oc apply -f https://raw.githubusercontent.com/openshift-knative/kn-plugin-func/tree/serverless-1.22.0/pipelines/resources/tekton/task/func-buildpacks/0.1/func-buildpacks.yaml
----

.. Create the `kn func` deploy Tekton task to be able to deploy the function in the pipeline:
+
[source,terminal]
----
$ oc apply -f https://raw.githubusercontent.com/openshift-knative/kn-plugin-func/tree/serverless-1.22.0/pipelines/resources/tekton/task/func-deploy/0.1/func-deploy.yaml
----

. Create a function:
+
[source,terminal]
----
$ kn func create <function_name> -l <runtime>
----

. Get a reference to the remote Git repository that hosts your function source code.

. Initialize a Git repository in your function project and add a reference to the remote repository.

> . Update the configuration in the `func.yaml` file for your function project to enable on cluster builds for the Git repository:

Reviewer 4: These steps are all a little too vague or incomplete, if we want to be thorough. After creating the function in the first step, the user needs to perform steps similar to the following.

* Initialize the project locally as a git repo
* Create a new repo in a tool such as github
* Use the commands provided by github UI to associate the project with the github repository
* Add the current project files to the repo with `git add`
* Commit that change with `git commit`
* Push those changes to github with `git push`

Then they can perform the on-cluster build.

It's sort of complicated and gets into areas that are really not ours to document. I wonder if instead of providing bulleted step by step procedures here, we use prose to describe the concepts. Would it be clearer and less prone to error?

> After creating a new function project with a command similar to that shown above, you'll need to add the project to a git repository and ensure that repository is available to the cluster. Then update `func.yaml` with your repository information as shown below.

Or something similar.

Author: Yeah I think that might be a good way to go
